cpp/src/arrow/util/bit_util.h

```diff
@@ -299,22 +307,22 @@ static constexpr bool GetBit(const uint8_t* bits, uint64_t i) {
 
 // Gets the i-th bit from a byte. Should only be used with i <= 7.
 static constexpr bool GetBitFromByte(uint8_t byte, uint8_t i) {
-  return byte & kBitmask[i];
+  return byte & GetBitMask(i);
 }
 
 static inline void ClearBit(uint8_t* bits, int64_t i) {
-  bits[i / 8] &= kFlippedBitmask[i % 8];
+  bits[i / 8] &= GetFlippedBitMask(i % 8);
 }
 
-static inline void SetBit(uint8_t* bits, int64_t i) { bits[i / 8] |= kBitmask[i % 8]; }
+static inline void SetBit(uint8_t* bits, int64_t i) { bits[i / 8] |= GetBitMask(i % 8); }
```

Comment:
One advantage that the `kBitmask` implementation had over this one is that memory access with negative `i` is UB, so the compiler was assuming here that `i >= 0` to come up with the bitmask.

We can bring that UB back (yes, UB is a Good Thing™️) by using `ARROW_COMPILER_ASSUME(i >= 0)`. The generated assembly for `SetBit` with the non-negative assumption is much shorter and doesn't need a conditional mov (or `csel` in ARM) instruction.

```asm
SetBit2(unsigned char*, long):                          # @SetBit2(unsigned char*, long)
        mov     rcx, rsi
        lea     rax, [rsi + 7]
        test    rsi, rsi
        cmovns  rax, rsi
        mov     edx, eax
        and     edx, 248
        sub     ecx, edx
        mov     edx, 1
        shl     edx, cl
        sar     rax, 3
        or      byte ptr [rdi + rax], dl
        ret
SetBit2NNeg(unsigned char*, long):                     # @SetBit2NNeg(unsigned char*, long)
        mov     ecx, esi
        and     cl, 7
        mov     al, 1
        shl     al, cl
        shr     rsi, 3
        or      byte ptr [rdi + rsi], al
        ret
```

All the experiments on Godbolt -> https://godbolt.org/z/Ez974vE3d